File: Model Branch/Condensatore.cpp
//
// Created by luca on 11/01/18.
//

#include "Condensatore.h"

Condensatore::Condensatore(double c):capacita(c){
  Componente::setImp(impedenza());
}
Condensatore::Condensatore(std::string s){
  auto pos=s.find('C');
  if(pos==0){
    capacita=std::stod(s.substr(1));
    Componente::setImp(impedenza());
  }
  else{
    std::cout << "errore di sintassi nella costruzione di condensatore" << std::endl;
    //------------------
    //GESTIRE ECCEZIONE
    //------------------
  }
}

Condensatore::Condensatore():capacita(0){};

bool Condensatore::operator== (const Dato& d) const{
    try {
        auto aux = dynamic_cast<const Condensatore &>(d);
        return capacita == aux.capacita;
    }
    catch(const std::bad_cast& error) {
        return false;
    }
}

Condensatore& Condensatore::operator= (const Dato& d){
    try{
        auto aux= dynamic_cast<const Condensatore&>(d);
        capacita = aux.capacita;
        setImp(capacita);
        return *this;
    }
    catch(const std::bad_cast& error){
        std::cout<<"tipi incompatibili"<<std::endl;
    }
}

C_cartesiano Condensatore::impedenza() const{
    double xc=1/(2*(Numero::pi)*(Componente::freq)*capacita);
    return C_cartesiano(0,-xc);
}

File: Java/Condensatore.java
// classe concreta Condensatore, derivata da Componente


// TO DO:
//   GESTIONE ECCEZIONI -> segnalate nell'implementazione dei vari metodi


//SCELTE PROGETTUALI:
// 1) override dei metodi toString(), equals(), impedenza()
// 3) "traduzione" in linguaggio Java delle ulteriori funzionalita' offerte dalla corrispettiva classe C++
//    a)costruttore da double
//    c)costruttore di default

class Condensatore extends Componente{

  private double capacita;

  public Condensatore(double d){
    capacita = d;
    super.setImp(impedenza());
  }

  public Condensatore(){
    capacita = 0;
  }

  public CCartesiano impedenza(){
    double aux = 1 / (2 * (Numero.pi) * (Componente.freq) * capacita);
    CCartesiano result = new CCartesiano(0, -aux);
    return result;
  }

  /*
  public void assign(Dato d){
    if(!(d instanceof Condensatore)){
      System.out.println("Condensatore.assign()-> d non e' instanza di Condensatore")
      //-------------------
      //GESTIRE ECCEZIONE
      //-------------------
    }

    Condensatore aux = (Condensatore) d;

    super.assign(aux);

    capacita = aux.capacita;
  }
  */

  public boolean equals(Object obj){
    if(!(obj instanceof Condensatore)){
      System.out.println("Condensatore.equals()-> obj non e' instanza di Condensatore");
      //-------------------
      //GESTIRE ECCEZIONE
      //-------------------
    }

    Condensatore aux = (Condensatore) obj;

    if(super.equals(aux)){
      if(capacita == aux.capacita){
        return true;
      }
      else{
        return false;
      }
    }
    else{
      return false;
    }
  }

  public String toString(){
    String result = super.toString();
    result = result + "CAPACITA': " + capacita + ";\r\n";
    return result;
  }
}
